xarray/coding/variables.py

```diff
@@ -280,8 +280,10 @@ def encode(self, variable: Variable, name: T_Name = None):
                     data = duck_array_ops.fillna(data, fill_value)
 
         if mv_exists:
+            # Use _FillValue if available to align missing_value to prevent issues
+            # when decoding
             # Ensure missing_value is cast to same dtype as data's
-            encoding["missing_value"] = dtype.type(mv)
+            encoding["missing_value"] = attrs.get("_FillValue", dtype.type(mv))
```

Comment:
We can't do this when `missing_value` is provided. They mean different things in netCDF land but are both interpreted as "NaN" in Xarray land.

Reply:
Yes, but if both are given, they need to be the same. This is already checked above and errors out. This here only aligns if they have different type float32 vs. float64, to common type.

Reply:
Oh thanks! Perhaps add a comment to be explicit about that.

Reply:
I've restructured the whole thing to make it more explicit what's going on. Also rephrased the comment. Not sure, if this is explicit enough.